master/pkg/stream/stream.go

```diff
@@ -183,36 +228,63 @@ func (p *Publisher[T]) Broadcast(events []Event[T]) {
 	wakeupID := p.WakeupID
 
 	// check each event against each subscription
-	for sub := range p.Subscriptions {
-		func() {
-			for _, ev := range events {
-				var msg interface{}
-				switch {
-				case ev.After != nil && sub.filter(*ev.After) && sub.permissionFilter(*ev.After):
```

Comment:
why do we need to change this logic at all. why can't we just pass `events []Event[T]` that are saturated? i was under the impression that is what we were going to do. i expected the diff now to just be a few lines in publisher.go like `satured, err := saturateEvents(events)` and then the rest of the code unchanged.

Reply:
1. This handles the case of :
events = [insert/fallin, delete]

If the last event is a deletion event, we are not able to saturate, so we can't send the message for the insert or fallin. Its ID is stored in `userNotKnownIDs` for skipping the deletion event, so the client doesn't have to worry about handling a non-existent deletion.

2. It handles the case of:
events = [update, update, update]

We are only able saturated message corresponding to the last update event or an event after the current batch if there is a queue. I use the logic below to avoid sending duplicate messages.
```
if cachedSeq == afterMsg.SeqNum() {
    msg = sub.Streamer.PrepareFn(recordCache.UpsertMsg)
```

3. If events = [update, fallout, update]

Since the saturate message is from the second update, we don't want to send message to the client when we process the first update event. The user shouldn't receive it since it's getting a fallout before the second update event. This is handle by:
```
if cachedSeq == afterMsg.SeqNum() {
    msg = sub.Streamer.PrepareFn(recordCache.UpsertMsg)
```

4. If events = [fallin/insert, fallout]

The saturate message is from fallout. We don't want to send the message when processing fallin/insert. This is also handle by:
```
if cachedSeq == afterMsg.SeqNum() {
    msg = sub.Streamer.PrepareFn(recordCache.UpsertMsg)
```
We also don't want to send a deletion message when we process the fallout event. I use `userNotKnownIDs` to store its ID during fallin/insert and no deletion message will be send when we process the fallout.

Reply:
Why are we not sending a deletion message when we process fallout? In the past we had treated a fallout as if it was a deletion, is that not the case anymore? My concern with not sending a deletion is the client will still believe it exists when it's no longer in scope.

Reply:
We are not sending deletion message when we process fallout in case 4, because the client doesn't have the message from fallin/insert event. Sorry for the confusion. When you are up for a chat, happy to about it. I'm not sure how to make this logic easier to understand.

Reply:
yeah exactly!